examples/audio/vocal_track_separation.py

```diff
@@ -0,0 +1,611 @@
+    first_conv_out = layers.Conv2D(channels, 1, 1, use_bias=False)(mix)
+    x = first_conv_out
+
+    # Encoder path
+    encoder_outs = []
+    for _ in range(n_scales):
+        x = tfc_tdf(x, channels, block_length, fft_dim, bottleneck_factor)
+        encoder_outs.append(x)
+        fft_dim, channels = fft_dim // scale[0], channels + growth
+        x = downscale(x, channels, scale)
+
+    # Bottleneck
+    x = tfc_tdf(x, channels, block_length, fft_dim, bottleneck_factor)
+
+    # Decoder path
+    for _ in range(n_scales):
+        fft_dim, channels = fft_dim * scale[0], channels - growth
+        x = ops.concatenate([upscale(x, channels, scale), encoder_outs.pop()], axis=-1)
+        x = tfc_tdf(x, channels, block_length, fft_dim, bottleneck_factor)
+
+    # Residual connection and final convolutions
+    x = ops.concatenate([mix, x * first_conv_out], axis=-1)
+    x = layers.Conv2D(channels, 1, 1, use_bias=False, activation="gelu")(x)
+    x = layers.Conv2D(n_instruments * n_subbands * 2, 1, 1, use_bias=False)(x)
+
+    # Reshape back to instrument-wise STFT
+    x = ops.reshape(x, (-1, x.shape[1], x.shape[2] * n_subbands, n_instruments, 2))
+    x = ops.transpose(x, (0, 3, 1, 2, 4))
+    x = ops.reshape(x, (-1, n_instruments, x.shape[2], x.shape[3] * 2))
+
+    return keras.Model(inputs=inputs, outputs=x, **kwargs)
+
+
+"""
+## Loss and Metrics
+
+We define:
+
+- `spectral_loss`: Mean absolute error in STFT domain.
+- `sdr`: Signal-to-Distortion Ratio, a common source separation metric.
+"""
+
+
+def prediction_to_wave(x, n_instruments=N_INSTRUMENTS):
+    """Convert STFT predictions back to waveform."""
+    x = ops.reshape(x, (-1, x.shape[2], x.shape[3] // 2, 2))
+    x = inverse_stft(x)
+    return ops.reshape(x, (-1, n_instruments, x.shape[1]))
+
+
+def target_to_stft(y):
+    """Convert target waveforms to their STFT representations."""
+    y = ops.reshape(y, (-1, CHUNK_SIZE))
+    y_real, y_imag = ops.stft(y, STFT_N_FFT, STFT_HOP_LENGTH, STFT_N_FFT)
+    y_real, y_imag = y_real[..., :-1], y_imag[..., :-1]
+    y = ops.stack([y_real, y_imag], axis=-1)
+    return ops.reshape(y, (-1, N_INSTRUMENTS, y.shape[1], y.shape[2] * 2))
+
+
+@saving.register_keras_serializable()
+def sdr(y_true, y_pred):
+    """Signal-to-Distortion Ratio metric."""
+    y_pred = prediction_to_wave(y_pred)
+    # Add epsilon for numerical stability
+    num = ops.sum(ops.square(y_true), axis=-1) + 1e-8
+    den = ops.sum(ops.square(y_true - y_pred), axis=-1) + 1e-8
+    return 10 * ops.log10(num / den)
+
+
+@saving.register_keras_serializable()
+def spectral_loss(y_true, y_pred):
+    """Mean absolute error in the STFT domain."""
+    y_true = target_to_stft(y_true)
+    return ops.mean(ops.absolute(y_true - y_pred))
+
+
+"""
+## Training
+"""
+
+# Load or create the model
+if path.exists(MODEL_PATH):
+    model = saving.load_model(MODEL_PATH)
+else:
+    model = tfc_tdf_net(keras.Input(sample_batch_x.shape[1:]), name="tfc_tdf_net")
+
+model.summary()
```

Comment:
Would it be useful to plot the model, or is the result too busy? e.g. `keras.utils.plot_model`

Reply:
This is how it would look that after a refactoring i've made grouping the TFC_TDF blocks.
It's a little long, but actually simple in structure. Regardless, It can be further refactored to group into decoder and encoder blocks, but perhaps it wont be as informative.

![model](https://github.com/user-attachments/assets/c79e842a-665e-4e91-8291-cc818d26c97e)

Reply:
Ok -- your call on whether to include it.